k4FWCore/components/IOSvc.cpp

```diff
@@ -0,0 +1,142 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "IOSvc.h"
+
+#include "edm4hep/MCParticleCollection.h"
+#include "podio/Frame.h"
+#include "podio/FrameCategories.h"
+
+#include "k4FWCore/KeepDropSwitch.h"
+
+#include "GaudiKernel/AnyDataWrapper.h"
+#include "GaudiKernel/IEventProcessor.h"
+
+#include <mutex>
+#include <thread>
+#include <tuple>
+
+StatusCode IOSvc::initialize() {
+  if (!m_readingFileNames.empty()) {
+    m_reader = std::make_unique<podio::ROOTReader>();
+    try {
+      m_reader->openFiles(m_readingFileNames);
+    } catch (std::runtime_error& e) {
+      error() << "Error when opening files: " << e.what() << endmsg;
+      throw e;
+    }
+    m_entries = m_reader->getEntries(podio::Category::Event);
+  }
+
+  m_switch = KeepDropSwitch(m_outputCommands);
+
+  m_incidentSvc = service("IncidentSvc");
+  m_incidentSvc->addListener(this, IncidentType::EndEvent);
+
+  m_dataSvc = service("EventDataSvc");
+
+  m_hiveWhiteBoard = service("EventDataSvc");
+
+  return Service::initialize();
+}
+
+StatusCode IOSvc::finalize() { return Service::finalize(); }
+
+std::tuple<std::vector<std::shared_ptr<podio::CollectionBase>>, std::vector<std::string>, podio::Frame> IOSvc::next() {
+  podio::Frame frame;
+  {
+    std::scoped_lock<std::mutex> lock(m_changeBufferLock);
+    info() << "m_nextEntry = " << m_nextEntry << " m_entries = " << m_entries << endmsg;
+    if (m_nextEntry < m_entries) {
+      frame = podio::Frame(std::move(m_reader->readEntry(podio::Category::Event, m_nextEntry)));
+    } else {
+      return std::make_tuple(std::vector<std::shared_ptr<podio::CollectionBase>>(), std::vector<std::string>(),
+                             std::move(frame));
+    }
+    m_nextEntry++;
+    if (m_collectionNames.empty()) {
+      m_collectionNames = frame.getAvailableCollections();
+    }
+  }
+
+  if (m_nextEntry >= m_entries) {
+    // if (true) {
+    auto       ep = serviceLocator()->as<IEventProcessor>();
+    StatusCode sc = ep->stopRun();
+    if (sc.isFailure()) {
+      error() << "Error when stopping run" << endmsg;
+      throw GaudiException("Error when stopping run", name(), StatusCode::FAILURE);
+    }
+    info() << "m_nextEntry = " << m_nextEntry << " m_entries = " << m_entries << endmsg;
+  }
+
+  std::vector<std::shared_ptr<podio::CollectionBase>> collections;
+
+  for (const auto& name : m_collectionNames) {
+    auto ptr = const_cast<podio::CollectionBase*>(frame.get(name));
+    collections.push_back(std::shared_ptr<podio::CollectionBase>(ptr));
+  }
+
+  return std::make_tuple(collections, m_collectionNames, std::move(frame));
+}
+
+// After every event if there is still a frame in the TES
+// that means it hasn't been written so the collections inside the Frame
+// should be removed so that they are deleted when the Frame is deleted
+void IOSvc::handle(const Incident& incident) {
```

Comment:
That's not needed, see https://github.com/key4hep/k4FWCore/blob/no-datasvc/k4FWCore/components/IOSvc.cpp#L49:
```
  m_incidentSvc->addListener(this, IncidentType::EndEvent);
```

Reply:
I'm aware that currently the service subscribe to only one type of incident, but the check isn't very expensive and it could potentially save some debugging time later. I'd say it also helps to understand the intentions and reflects what is written in the comment

Reply:
I don't see in which situation that would be needed, it's a callback for only that event; in case of it not working it's a bug in Gaudi and I imagine there are checks for that there.

Reply:
I don't insist
In the incident API it's impossible to specify different callbacks for different incident types -> the same method `handle` is called for every case, so usually the `handle` just checks the incident type and dispatches correct function for that incident.  For one incident type it seems like an overkill but illustrates intended usage

I wanted to be a bit pedantic and check because the API in principle allows the incident to be different  and potentially allows also to call `addListener` from outside. In normal scenario it's very unlikely to happen